File: src/dynamic_programming/cpp/can construct/src/can_construct.cpp
#include "can_construct.h"
#include <functional>
#include <unordered_map>

bool canConstructBasic(const std::string &target,
                       const std::vector<std::string> &wordBank) {

  if (target.empty())
    return true;

  for (const auto &word : wordBank) {

    if (target.size() >= word.size() && target.substr(0, word.size()) == word) {
      auto remainder = target.substr(word.size());

      if (canConstructBasic(remainder, wordBank))
        return true;
    }
  }

  return false;
}

bool canConstructMemo(const std::string &target,
                      const std::vector<std::string> &wordBank) {

  std::function<bool(const std::string &, const std::vector<std::string> &,
                     std::unordered_map<std::string, bool> &)>
      _canConstructMemo;
  _canConstructMemo = [&](const std::string &target,
                          const std::vector<std::string> &wordBank,
                          std::unordered_map<std::string, bool> &memo) -> bool {
    if (target.empty())
      return true;

    if (memo.count(target))
      return memo[target];

    for (const auto &word : wordBank) {

      if (target.size() >= word.size() &&
          target.substr(0, word.size()) == word) {
        auto remainder = target.substr(word.size());

        if (_canConstructMemo(remainder, wordBank, memo)) {
          memo[target] = true;
          return true;
        }
      }
    }

    memo[target] = false;
    return false;
  };

  std::unordered_map<std::string, bool> memo;

  return _canConstructMemo(target, wordBank, memo);
}

bool canConstructTable(const std::string &target,
                       const std::vector<std::string> &wordBank) {

  std::vector<bool> table(target.size() + 1, false);
  table[0] = true;

  for (int i = 0; i < target.size(); i++) {

    if (table[i]) {
      for (const auto &word : wordBank) {

        if (target.substr(i, word.size()) == word) {

          table[i + word.size()] = true;
        }
      }
    }
  }

  return table.back();
}

File: src/dynamic_programming/python/can_construct/src/can_construct.py
from typing import List, Optional


def can_construct_basic(target: str, word_bank: List[str]) -> bool:
    """
    Determine if target string can be constructed from words in word_bank.

    Uses basic recursion without memoization.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).

    Returns:
        True if target can be constructed, False otherwise.
    """
    if not target:
        return True

    return any(
        len(target) >= len(word)
        and target[: len(word)] == word
        and can_construct_basic(target[len(word) :], word_bank)
        for word in word_bank
    )


def can_construct_memo(
    target: str, word_bank: List[str], memo: Optional[dict] = None
) -> bool:
    """
    Determine if target string can be constructed from words in word_bank.

    Uses memoization for efficient computation.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).
        memo: Optional dictionary for memoization cache.

    Returns:
        True if target can be constructed, False otherwise.
    """
    if memo is None:
        memo = {}

    if not target:
        return True

    if target in memo:
        return memo[target]

    can_construct = any(
        len(target) >= len(word)
        and target[: len(word)] == word
        and can_construct_memo(target[len(word) :], word_bank, memo)
        for word in word_bank
    )
    memo[target] = can_construct

    return can_construct


def can_construct_table(target: str, word_bank: List[str]) -> bool:
    """
    Determine if target string can be constructed from words in word_bank.

    Uses tabulation (bottom-up DP) for efficient computation.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).

    Returns:
        True if target can be constructed, False otherwise.
    """
    table = [False] * (len(target) + 1)
    table[0] = True

    for i in range(len(target)):
        if table[i]:
            for word in word_bank:
                if target[i : i + len(word)] == word:
                    table[i + len(word)] = True

    return table[-1]
